extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/CoroutineInvocationHandler.kt

```diff
@@ -1,25 +1,43 @@
 package org.jboss.resteasy.reactive.server.runtime.kotlin
 
-import kotlinx.coroutines.GlobalScope
+import io.vertx.core.Vertx
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.launch
+import org.eclipse.microprofile.context.ManagedExecutor
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext
 import org.jboss.resteasy.reactive.server.spi.EndpointInvoker
 import org.jboss.resteasy.reactive.server.spi.ServerRestHandler
+import org.slf4j.LoggerFactory
+
+private val logger = LoggerFactory.getLogger(CoroutineInvocationHandler::class.java)
+
+class CoroutineInvocationHandler(private val invoker: EndpointInvoker,
+                                 private val coroutineScope: CoroutineScope,
+                                 managedExecutor: ManagedExecutor) : ServerRestHandler {
+
+    private val executorDispatcher = ExecutorDispatcher(managedExecutor)
 
-class CoroutineInvocationHandler(private val invoker: EndpointInvoker) : ServerRestHandler {
     override fun handle(requestContext: ResteasyReactiveRequestContext) {
         if (requestContext.result != null) {
             return
         }
+        if (invoker !is CoroutineEndpointInvoker) {
+            requestContext.handleException(IllegalStateException("Not a coroutine invoker"), true)
+            return
+        }
+
+        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher)
+                // The @Blocking annotation will not run in a Vertx context
+                ?: executorDispatcher
```

Comment:
How about [asCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/java.util.concurrent.-executor/as-coroutine-dispatcher.html)?

```kotlin
        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher)
                // The @Blocking annotation will not run in a Vertx context
                ?: managedExecutor.asCoroutineDispatcher()
```

Reply:
The wrapper dispatcher around executor services is not deterministic. If the executor service rejects the task, it gets submitted to `Dispatchers.IO`.
I think is best to keep it deterministic and if the task submission fails, for instance, the executor service is being shut down, then it will fail the request.

Reply:
Your're right, nice catch 👍.

Reply:
After reading more about [asCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/java.util.concurrent.-executor/as-coroutine-dispatcher.html), i'm not sure anymore.

It is correkt, that the task is submitted to the [Dispatchers.IO](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-i-o.html), but only `... to cleanup its resources and promptly complete` ([Code](https://github.com/Kotlin/kotlinx.coroutines/blob/32336be8fa90818cff9f2ba3c36b3f7c89e079c4/kotlinx-coroutines-core/jvm/src/Executors.kt#L92-L100)).

It might be better to use [asCoroutineDispatcher](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/java.util.concurrent.-executor/as-coroutine-dispatcher.html) in this case, because it is an `Executor` :thinking:

Reply:
Don't see the advantage of being an `Executor` per se.

Regarding the resource cleanup it makes sense. I'm just wondering If I want to run additional code when the Quarkus runtime already decided to shutdown its executors.
Would like to have feedback from other reviewers on this one.

Reply:
Since we didn't had any additional feedback and the latest discussions removed the executor, I'll resolve this as well.